File: apps/api/app/schemas/product.py
# schemas/product.py

from pydantic import BaseModel, Field, ConfigDict, computed_field
from typing import Optional, List, Dict, Any
from datetime import datetime

# ==========================================
# LEGACY SCHEMAS (KEEP FOR COMPATIBILITY)
# ==========================================

class ProductSchemaLegacy(BaseModel):
    """Original schema - keep for backward compatibility"""
    id: int
    name: str
    description: Optional[str] = None
    price: float  # Keep as float (dollars) for legacy compatibility
    image_url: Optional[str]
    image_urls: Optional[List[str]] = []
    category: Optional[str] = None  # Keep as string
    featured: Optional[bool] = None
    details: Optional[Dict[str, str]] = {}
    display_theme: Optional[str] = "dark"

    class Config:
        from_attributes = True

class PaginatedProductsResponseLegacy(BaseModel):
    """Original paginated response - keep for compatibility"""
    products: List[ProductSchemaLegacy]
    total: int
    page: int
    pageSize: int
    totalPages: int

    class Config:
        from_attributes = True

# ==========================================
# ENHANCED SCHEMAS (MATCH YOUR DATABASE)
# ==========================================

class CategoryBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    slug: Optional[str] = None
    description: Optional[str] = None
    meta_title: Optional[str] = Field(None, max_length=200)
    meta_description: Optional[str] = Field(None, max_length=500)  # 500 in your DB
    parent_id: Optional[int] = None
    sort_order: int = Field(default=0)
    is_active: bool = Field(default=True)
    is_featured: bool = Field(default=False)

class CategoryOut(CategoryBase):
    id: int
    full_path: Optional[str] = None
    children_count: int = Field(default=0)
    products_count: int = Field(default=0)
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    model_config = ConfigDict(from_attributes=True)

class CollectionBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    slug: Optional[str] = None
    description: Optional[str] = None
    collection_type: str = Field(default="manual")
    is_active: bool = Field(default=True)
    is_featured: bool = Field(default=False)

class CollectionOut(CollectionBase):
    id: int
    products_count: int = Field(default=0)
    created_at: datetime
    
    model_config = ConfigDict(from_attributes=True)

class ProductBase(BaseModel):
    """Base product schema matching your database fields"""
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    price: int = Field(..., gt=0)  # Price in cents (matches your DB)
    
    # Product identification
    sku: Optional[str] = Field(None, max_length=50)
    slug: Optional[str] = Field(None, max_length=200)
    short_description: Optional[str] = Field(None, max_length=500)
    
    # Pricing
    compare_at_price: Optional[int] = None
    cost_price: Optional[int] = None
    
    # Categories
    category_id: Optional[int] = None
    category: Optional[str] = None  # Legacy field
    
    # Inventory - FIXED field names to match your DB
    inventory_count: Optional[int] = Field(default=0)
    track_inventory: Optional[bool] = Field(default=True)
    inventory_policy: Optional[str] = Field(default="deny")
    low_stock_threshold: Optional[int] = None
    
    # Product attributes
    weight: Optional[float] = None
    dimensions: Optional[Dict[str, Any]] = None  # JSONB in DB
    materials: Optional[List[str]] = None
    gemstones: Optional[Dict[str, Any]] = None  # JSONB in DB, not List
    care_instructions: Optional[str] = None
    product_type: Optional[str] = Field(None, max_length=50)
    
    # Media - FIXED field names to match your DB
    image_url: Optional[str] = None
    image_urls: Optional[List[str]] = None
    featured_image: Optional[str] = None
    video_url: Optional[str] = None
    image_alt_texts: Optional[List[str]] = None  # NOT alt_texts
    
    # Status and visibility
    status: Optional[str] = Field(default="active")
    featured: bool = Field(default=False)
    searchable: Optional[bool] = Field(default=True)
    available_online: Optional[bool] = Field(default=True)
    available_in_store: Optional[bool] = Field(default=True)
    
    # SEO - FIXED field names to match your DB
    meta_title: Optional[str] = Field(None, max_length=200)
    meta_description: Optional[str] = Field(None, max_length=500)  # 500 in your DB
    search_keywords: Optional[List[str]] = None  # NOT meta_keywords
    social_share_title: Optional[str] = Field(None, max_length=200)
    social_share_description: Optional[str] = Field(None, max_length=500)
    
    # Business intelligence
    view_count: Optional[int] = Field(default=0)
    conversion_rate: Optional[float] = Field(default=0.0)
    average_rating: Optional[float] = Field(default=0.0)
    review_count: Optional[int] = Field(default=0)
    
    # Legacy fields
    details: Optional[Dict[str, Any]] = {}
    display_theme: Optional[str] = "dark"
    
    # Admin fields
    admin_notes: Optional[str] = None
    created_by: Optional[str] = None
    last_modified_by: Optional[str] = None

class ProductOut(ProductBase):
    """Enhanced product schema for API responses - Let Pydantic handle it"""
    id: int
    
    # Category handling - FIXED for both legacy and new
    category_string: Optional[str] = Field(None, alias="category")  # Legacy string field
    category_id: Optional[int] = None
    product_category: Optional[CategoryOut] = None  # New Category object
    
    # Collections relationship
    collections: Optional[List[CollectionOut]] = None
    
    # Timestamps
    created_at: datetime
    updated_at: Optional[datetime] = None
    published_at: Optional[datetime] = None
    discontinued_at: Optional[datetime] = None
    
    model_config = ConfigDict(from_attributes=True)
    
    # Computed fields - Let Pydantic calculate these
    @computed_field
    @property
    def price_display(self) -> str:
        """Format price for display (convert cents to dollars)"""
        return f"${(self.price or 0) / 100:.2f}"
    
    @computed_field
    @property
    def compare_price_display(self) -> Optional[str]:
        """Format compare price for display"""
        return f"${(self.compare_at_price or 0) / 100:.2f}" if self.compare_at_price else None
    
    @computed_field
    @property
    def is_on_sale(self) -> bool:
        """Check if product is on sale"""
        return bool(self.compare_at_price and self.price and self.price < self.compare_at_price)
    
    @computed_field
    @property
    def discount_percentage(self) -> int:
        """Calculate discount percentage"""
        if not self.is_on_sale or not self.compare_at_price:
            return 0
        return int(((self.compare_at_price - self.price) / self.compare_at_price) * 100)
    
    @computed_field
    @property
    def in_stock(self) -> bool:
        """Check if product is available for purchase"""
        if not self.track_inventory:
            return True
        return (self.inventory_count or 0) > 0
    
    @computed_field
    @property
    def category_name(self) -> Optional[str]:
        """Get category name from either source"""
        if self.product_category:
            return self.product_category.name
        return self.category_string

class ProductCreate(BaseModel):
    """Schema for creating products - required fields only"""
    name: str = Field(..., min_length=1, max_length=200)
    price: int = Field(..., gt=0)  # Required
    description: Optional[str] = None
    category_id: Optional[int] = None
    sku: Optional[str] = Field(None, max_length=50)
    image_url: Optional[str] = None
    featured: bool = Field(default=False)
    status: str = Field(default="active")

class ProductUpdate(BaseModel):
    """Schema for updating products - all optional"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    price: Optional[int] = Field(None, gt=0)
    sku: Optional[str] = Field(None, max_length=50)
    slug: Optional[str] = None
    category_id: Optional[int] = None
    inventory_count: Optional[int] = None
    track_inventory: Optional[bool] = None
    weight: Optional[float] = None
    materials: Optional[List[str]] = None
    image_url: Optional[str] = None
    image_urls: Optional[List[str]] = None
    status: Optional[str] = None
    featured: Optional[bool] = None
    details: Optional[Dict[str, Any]] = None
    display_theme: Optional[str] = None

class ProductSummary(BaseModel):
    """Lightweight schema for product lists"""
    id: int
    name: str
    price: int  # Cents
    price_display: str  # $24.99
    image_url: Optional[str] = None
    slug: Optional[str] = None
    category_name: Optional[str] = None
    featured: bool = Field(default=False)
    in_stock: bool = Field(default=True)
    average_rating: Optional[float] = Field(default=0.0)
    display_theme: Optional[str] = "dark"
    
    model_config = ConfigDict(from_attributes=True)

class ProductListResponse(BaseModel):
    """New paginated response format"""
    products: List[ProductSummary]
    total: int
    page: int
    page_size: int
    total_pages: int
    has_next: bool
    has_prev: bool

# ==========================================
# CONVERSION HELPERS (UPDATED)
# ==========================================

def convert_product_to_legacy(product) -> ProductSchemaLegacy:
    """Convert enhanced product to legacy format"""
    return ProductSchemaLegacy(
        id=product.id,
        name=product.name,
        description=product.description,
        price=(product.price / 100) if product.price else 0,  # Convert cents to dollars
        image_url=product.image_url,
        image_urls=product.image_urls or [],
        category=getattr(product.category_obj, 'name', None) if hasattr(product, 'category_obj') else product.category,
        featured=product.featured,
        details=product.details or {},
        display_theme=product.display_theme
    )

def create_product_summary(product) -> ProductSummary:
    """Create ProductSummary from Product model"""
    return ProductSummary(
        id=product.id,
        name=product.name,
        price=product.price or 0,
        price_display=f"${(product.price / 100):.2f}" if product.price else "$0.00",
        image_url=product.image_url,
        slug=product.slug,
        category_name=getattr(product.category_obj, 'name', None) if hasattr(product, 'category_obj') else product.category,
        featured=product.featured or False,
        in_stock=product.is_in_stock if hasattr(product, 'is_in_stock') else True,
        average_rating=product.average_rating or 0.0
    )

File: apps/web/src/types/product.ts
// types/product.ts - CORRECTED to match backend Epic #11 schema

// ==========================================
// JSON/JSONB FIELD TYPES
// ==========================================

export interface ProductDimensions {
    length?: number;
    width?: number;
    height?: number;
    unit?: string;
}

export interface GemstoneInfo {
    type: string;
    cut?: string;
    carat?: number;
    color?: string;
    clarity?: string;
    origin?: string;
    certification?: string;
    [key: string]: string | number | undefined;
}

export interface ProductDetails {
    style?: string;
    occasion?: string;
    collection?: string;
    limited_edition?: boolean;
    warranty_years?: number;
    [key: string]: string | number | boolean | undefined;
}

// ==========================================
// CORE PRODUCT TYPES
// ==========================================

export interface Product {
    // Core fields - MATCH ProductSummary from backend
    id: number;
    name: string;
    description?: string;
    price: number; // Price in cents (matches backend)
    price_display: string; // Formatted price display "$24.99"

    // Product identification
    sku?: string;
    slug?: string;
    short_description?: string; // Added to match backend

    // Pricing - ADDED to match backend
    compare_at_price?: number; // In cents
    cost_price?: number; // In cents

    // Relationships
    category_id?: number;
    category?: Category; // Will be populated from product_category relationship
    category_name?: string; // From ProductSummary
    collections?: Collection[];

    // Inventory - CORRECTED field names to match backend
    inventory_count: number;
    inventory_policy?: 'deny' | 'continue';
    track_inventory?: boolean;
    low_stock_threshold?: number;
    in_stock: boolean; // Computed field

    // Product attributes - CORRECTED types to match backend
    weight?: number;
    dimensions?: ProductDimensions; // JSONB in backend
    materials?: string[]; // ARRAY(String) in backend
    gemstones?: GemstoneInfo; // ✅ FIXED: JSONB in backend with proper typing
    care_instructions?: string;
    product_type?: string;

    // Media - CORRECTED field names to match backend
    image_url?: string;
    image_urls?: string[];
    featured_image?: string;
    image_alt_texts?: string[]; // FIXED: backend uses image_alt_texts, not alt_texts
    video_url?: string;

    // SEO - CORRECTED field names to match backend
    meta_title?: string;
    meta_description?: string;
    search_keywords?: string[]; // FIXED: backend uses search_keywords, not meta_keywords
    social_share_title?: string; // Added to match backend
    social_share_description?: string; // Added to match backend

    // Status
    status?: 'active' | 'draft' | 'archived';
    featured: boolean;
    searchable?: boolean; // Added to match backend
    available_online?: boolean; // Added to match backend
    available_in_store?: boolean; // Added to match backend

    // Business intelligence
    view_count?: number;
    conversion_rate?: number; // Added to match backend
    average_rating?: number;
    review_count?: number;

    // Legacy fields - Keep for compatibility
    details?: ProductDetails; // ✅ FIXED: JSONB in backend with proper typing
    display_theme?: string;

    // Admin
    admin_notes?: string;
    created_by?: string;
    last_modified_by?: string; // Added to match backend

    // Timestamps
    created_at: string;
    updated_at?: string;
    published_at?: string; // Added to match backend
    discontinued_at?: string; // Added to match backend
}

export interface Category {
    id: number;
    name: string;
    slug?: string;
    description?: string;

    // SEO
    meta_title?: string;
    meta_description?: string;

    // Hierarchy - ADDED to match backend
    parent_id?: number;
    level?: number; // Added to match backend
    sort_order: number;
    full_path?: string; // For breadcrumbs: "Jewelry > Rings > Engagement"

    // Display - ADDED to match backend
    image_url?: string;
    banner_image?: string;
    icon_class?: string;

    // Computed fields
    children_count: number;
    products_count: number;

    // Status
    is_active: boolean;
    is_featured: boolean;
    is_menu_visible?: boolean; // Added to match backend

    // Admin - ADDED to match backend
    created_by?: string;
    admin_notes?: string;

    // Timestamps
    created_at: string;
    updated_at?: string;
}

export interface Collection {
    id: number;
    name: string;
    slug?: string;
    description?: string;

    // Collection type - CORRECTED to match backend
    collection_type: 'manual' | 'smart' | 'seasonal'; // Updated types
    purpose?: string; // Added to match backend
    smart_rules?: string; // Text field in backend, not Record

    // Visual and marketing - ADDED to match backend
    image_url?: string;
    banner_image?: string;
    video_url?: string;
    color_theme?: string; // Renamed from theme_color
    marketing_copy?: string;

    // Access control - CORRECTED to match backend
    is_public?: boolean; // Boolean in backend
    requires_password?: boolean; // Boolean in backend
    access_password?: string;

    // Campaign tracking - ADDED to match backend
    campaign_start?: string;
    campaign_end?: string;
    view_count: number;
    conversion_rate: number;

    // SEO
    meta_title?: string;
    meta_description?: string;

    // Computed/display
    products_count: number;
    sort_order: number;
    products_sort_by?: string; // Added to match backend

    // Status
    is_active: boolean;
    is_featured: boolean;

    // Admin - ADDED to match backend
    created_by?: string;
    admin_notes?: string;

    // Timestamps
    created_at: string;
    updated_at?: string;
}

export interface ProductListResponse {
    products: Product[]; // These are actually ProductSummary objects
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
    has_next: boolean;
    has_prev: boolean;
}

export interface CategoryListResponse {
    categories: Category[];
    total: number;
}

export interface CollectionListResponse {
    collections: Collection[];
    total: number;
}

// ==========================================
// FILTER TYPES - CORRECTED
// ==========================================

export interface ProductFilters {
    // Search
    name?: string;

    // Category filtering
    category_id?: number;
    category?: string; // Legacy support

    // Collection filtering
    collection_id?: number;

    // Price filtering (in dollars - will be converted to cents)
    minPrice?: number;
    maxPrice?: number;

    // Status filtering
    featured?: boolean;
    status?: 'active' | 'draft' | 'archived';
    in_stock?: boolean;

    // Pagination
    page?: number;
    pageSize?: number;

    // Sorting
    sortBy?: 'name' | 'price' | 'created_at' | 'view_count' | 'average_rating';
    sortOrder?: 'asc' | 'desc';
}

export interface CategoryFilters {
    parent_id?: number;
    include_inactive?: boolean;
    featured_only?: boolean;
}

export interface CollectionFilters {
    collection_type?: 'manual' | 'smart' | 'seasonal';
    featured_only?: boolean;
    include_inactive?: boolean;
}

// ==========================================
// FORM TYPES - CORRECTED
// ==========================================

export interface ProductCreateForm {
    name: string;
    description?: string;
    price: number; // In dollars (will be converted to cents)
    category_id: number;
    sku?: string;
    slug?: string;

    // Inventory
    inventory_count: number;
    inventory_policy: 'deny' | 'continue';
    track_inventory: boolean;

    // Attributes
    weight?: number;
    dimensions?: ProductDimensions;
    materials?: string[];
    gemstones?: GemstoneInfo; // ✅ FIXED: Proper typing instead of any
    care_instructions?: string;

    // Media
    image_url?: string;
    image_urls?: string[];
    image_alt_texts?: string[]; // FIXED: field name
    video_url?: string;

    // SEO
    meta_title?: string;
    meta_description?: string;
    search_keywords?: string[]; // FIXED: field name

    // Status
    status: 'active' | 'draft' | 'archived';
    featured: boolean;

    // Admin
    admin_notes?: string;
}

export interface CategoryCreateForm {
    name: string;
    slug?: string;
    description?: string;
    meta_title?: string;
    meta_description?: string;
    parent_id?: number;
    sort_order: number;
    is_active: boolean;
    is_featured: boolean;
}

export interface CollectionCreateForm {
    name: string;
    slug?: string;
    description?: string;
    collection_type: 'manual' | 'smart' | 'seasonal';
    smart_rules?: string; // Text field
    marketing_copy?: string;
    is_public: boolean;
    requires_password: boolean;
    access_password?: string;
    color_theme?: string;
    campaign_start?: string;
    campaign_end?: string;
    meta_title?: string;
    meta_description?: string;
    sort_order: number;
    is_active: boolean;
    is_featured: boolean;
}

// ==========================================
// COMPONENT PROP TYPES
// ==========================================

export interface ProductCardProps {
    product: Product;
    showPrice?: boolean;
    showCategory?: boolean;
    showRating?: boolean;
    index?: number;
    className?: string;
    disableRotation?: boolean;
    onClick?: (product: Product) => void;
}

export interface CategoryCardProps {
    category: Category;
    showProductCount?: boolean;
    showDescription?: boolean;
    className?: string;
    onClick?: (category: Category) => void;
}

export interface CollectionCardProps {
    collection: Collection;
    showProductCount?: boolean;
    showDescription?: boolean;
    className?: string;
    onClick?: (collection: Collection) => void;
}

export interface ProductGridProps {
    products: Product[];
    loading?: boolean;
    error?: string | null;
    showPrices?: boolean;
    showCategories?: boolean;
    className?: string;
    onProductClick?: (product: Product) => void;
}

export interface CategoryGridProps {
    categories: Category[];
    loading?: boolean;
    error?: string | null;
    showProductCounts?: boolean;
    className?: string;
    onCategoryClick?: (category: Category) => void;
}

export interface ProductFiltersProps {
    filters: ProductFilters;
    categories: Category[];
    collections: Collection[];
    onFiltersChange: (filters: Partial<ProductFilters>) => void;
    onClearFilters: () => void;
    totalResults?: number;
    loading?: boolean;
}

// ==========================================
// LEGACY COMPATIBILITY TYPES
// ==========================================

export interface ProductLegacy {
    id: number;
    name: string;
    description?: string;
    price: number; // In dollars
    image_url?: string;
    category?: string; // Category name as string
    featured: boolean;
}

export interface ProductListResponseLegacy {
    products: ProductLegacy[];
    total: number;
    page: number;
    pageSize: number;
}

// ==========================================
// UTILITY TYPES
// ==========================================

export type ProductStatus = 'active' | 'draft' | 'archived';
export type InventoryPolicy = 'deny' | 'continue';
export type CollectionType = 'manual' | 'smart' | 'seasonal';
export type SortOrder = 'asc' | 'desc';
export type SortField = 'name' | 'price' | 'created_at' | 'view_count' | 'average_rating';

// ==========================================
// TYPE GUARDS - ✅ FIXED: Proper typing instead of any
// ==========================================

export const isProduct = (item: unknown): item is Product => {
    return Boolean(
        item &&
        typeof item === 'object' &&
        'id' in item &&
        'name' in item &&
        'price' in item &&
        typeof (item as Product).id === 'number' &&
        typeof (item as Product).name === 'string' &&
        typeof (item as Product).price === 'number'
    );
};

export const isCategory = (item: unknown): item is Category => {
    return Boolean(
        item &&
        typeof item === 'object' &&
        'id' in item &&
        'name' in item &&
        typeof (item as Category).id === 'number' &&
        typeof (item as Category).name === 'string'
    );
};

export const isCollection = (item: unknown): item is Collection => {
    return Boolean(
        item &&
        typeof item === 'object' &&
        'id' in item &&
        'name' in item &&
        'collection_type' in item &&
        typeof (item as Collection).id === 'number' &&
        typeof (item as Collection).name === 'string' &&
        typeof (item as Collection).collection_type === 'string'
    );
};

// ==========================================
// CONSTANTS
// ==========================================

export const PRODUCT_STATUSES = ['active', 'draft', 'archived'] as const;
export const INVENTORY_POLICIES = ['deny', 'continue'] as const;
export const COLLECTION_TYPES = ['manual', 'smart', 'seasonal'] as const;

export const SORT_OPTIONS = [
    { value: 'name', label: 'Name' },
    { value: 'price', label: 'Price' },
    { value: 'created_at', label: 'Date Added' },
    { value: 'view_count', label: 'Popularity' },
    { value: 'average_rating', label: 'Rating' }
] as const;

export const JEWELRY_MATERIALS = [
    'Gold', '14K Gold', '18K Gold', 'White Gold', 'Rose Gold',
    'Silver', 'Sterling Silver', 'Platinum', 'Titanium',
    'Stainless Steel', 'Copper', 'Bronze'
] as const;